File: shared/did.py
"""did:key (Ed25519) identity primitives, per spec/auth.md and spec/patterns.md.

A did:key IS an Ed25519 public key: multicodec ed25519-pub (0xed 0x01) then the 32
raw bytes, base58btc-encoded with a 'z' multibase prefix, so it starts did:key:z6Mk.
Nothing is registered anywhere; resolution is offline and this file is the resolver.

Wire signature encoding is base64url, unpadded, 86 chars for a 64-byte signature
(auth.md). We never store or transmit a private key from this module; callers hold
the Ed25519PrivateKey object and pass it in to sign.

The DID note lives at /kv/did-<shard>/<key>, where the fingerprint is the first 16
hex of SHA-256 of the full did:key string, shard is its first 2 chars, key the rest.
"""
import base64
import hashlib

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

ED25519_MULTICODEC = b"\xed\x01"
_B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_B58_INDEX = {c: i for i, c in enumerate(_B58)}
_DID_PREFIX = "did:key:z"


def _b58encode(b):
    n = int.from_bytes(b, "big")
    out = ""
    while n > 0:
        n, r = divmod(n, 58)
        out = _B58[r] + out
    pad = 0
    for ch in b:
        if ch == 0:
            pad += 1
        else:
            break
    return "1" * pad + out


def _b58decode(s):
    n = 0
    for ch in s:
        if ch not in _B58_INDEX:
            raise ValueError("not base58btc: %r" % ch)
        n = n * 58 + _B58_INDEX[ch]
    body = n.to_bytes((n.bit_length() + 7) // 8, "big") if n else b""
    pad = 0
    for ch in s:
        if ch == "1":
            pad += 1
        else:
            break
    return b"\x00" * pad + body


def generate():
    """A fresh Ed25519 keypair. Returns (private_key_obj, did_str)."""
    priv = Ed25519PrivateKey.generate()
    return priv, did_from_priv(priv)


def pub_raw(priv):
    return priv.public_key().public_bytes_raw()


def did_from_pub_raw(pub):
    if len(pub) != 32:
        raise ValueError("ed25519 public key must be 32 bytes, got %d" % len(pub))
    return _DID_PREFIX + _b58encode(ED25519_MULTICODEC + pub)


def did_from_priv(priv):
    return did_from_pub_raw(pub_raw(priv))


def pub_raw_from_did(did):
    if not isinstance(did, str) or not did.startswith(_DID_PREFIX):
        raise ValueError("not a did:key:z... string")
    raw = _b58decode(did[len(_DID_PREFIX):])
    if raw[:2] != ED25519_MULTICODEC:
        raise ValueError("did:key is not ed25519-pub multicodec")
    pub = raw[2:]
    if len(pub) != 32:
        raise ValueError("decoded ed25519 key is not 32 bytes")
    return pub


def sign_b64url(priv, message):
    """Sign message (bytes) and return the 86-char unpadded base64url the wire wants."""
    return base64.urlsafe_b64encode(priv.sign(message)).rstrip(b"=").decode("ascii")


def verify_b64url(pub, sig_b64url, message):
    """True iff sig_b64url is a valid Ed25519 signature by pub (32 raw bytes) over
    message. Never raises: a malformed signature or key is a False, not a crash."""
    try:
        pad = "=" * (-len(sig_b64url) % 4)
        sig = base64.urlsafe_b64decode(sig_b64url + pad)
        Ed25519PublicKey.from_public_bytes(pub).verify(sig, message)
        return True
    except Exception:
        return False


def verify_by_did(did, sig_b64url, message):
    """verify_b64url resolving the key from the DID string itself."""
    try:
        return verify_b64url(pub_raw_from_did(did), sig_b64url, message)
    except Exception:
        return False


def fingerprint(did):
    """First 16 hex of SHA-256 of the full did:key string, lowercase (patterns.md)."""
    return hashlib.sha256(did.encode("utf-8")).hexdigest()[:16]


def note_shard_key(did):
    fp = fingerprint(did)
    return fp[:2], fp[2:]


def did_note_ns(did):
    """The note namespace for this DID: 'did-<shard>'."""
    shard, _ = note_shard_key(did)
    return "did-" + shard

File: agent/src/shared/did.ts
import { b64urlToBytes, bytesEqual, bytesToB64url, hexToBytes, sha256Hex } from "./bytes";

const PKCS8_PREFIX_HEX = "302e020100300506032b657004220420";

const B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
const B58_INDEX: Record<string, number> = {};
for (let i = 0; i < B58.length; i++) B58_INDEX[B58[i]!] = i;

const ED25519_MULTICODEC = Uint8Array.from([0xed, 0x01]);
const DID_PREFIX = "did:key:z";

function b58encode(b: Uint8Array): string {
  let n = 0n;
  for (const byte of b) n = n * 256n + BigInt(byte);
  let out = "";
  while (n > 0n) {
    const r = Number(n % 58n);
    out = B58[r]! + out;
    n = n / 58n;
  }
  let pad = 0;
  for (const byte of b) {
    if (byte === 0) pad++;
    else break;
  }
  return "1".repeat(pad) + out;
}

function b58decode(s: string): Uint8Array {
  let n = 0n;
  for (const ch of s) {
    const i = B58_INDEX[ch];
    if (i === undefined) throw new Error(`not base58btc: ${ch}`);
    n = n * 58n + BigInt(i);
  }
  const body: number[] = [];
  while (n > 0n) {
    body.unshift(Number(n & 0xffn));
    n = n >> 8n;
  }
  let pad = 0;
  for (const ch of s) {
    if (ch === "1") pad++;
    else break;
  }
  const out = new Uint8Array(pad + body.length);
  out.set(body, pad);
  return out;
}

export function didFromPubRaw(pub: Uint8Array): string {
  if (pub.length !== 32) throw new Error(`ed25519 public key must be 32 bytes, got ${pub.length}`);
  const payload = new Uint8Array(ED25519_MULTICODEC.length + pub.length);
  payload.set(ED25519_MULTICODEC, 0);
  payload.set(pub, ED25519_MULTICODEC.length);
  return DID_PREFIX + b58encode(payload);
}

export function pubRawFromDid(did: string): Uint8Array {
  if (typeof did !== "string" || !did.startsWith(DID_PREFIX)) {
    throw new Error("not a did:key:z... string");
  }
  const raw = b58decode(did.slice(DID_PREFIX.length));
  if (raw.length < 2 || raw[0] !== ED25519_MULTICODEC[0] || raw[1] !== ED25519_MULTICODEC[1]) {
    throw new Error("did:key is not ed25519-pub multicodec");
  }
  const pub = raw.slice(2);
  if (pub.length !== 32) throw new Error("decoded ed25519 key is not 32 bytes");
  return pub;
}

export function pubRawEquals(did: string, pub: Uint8Array): boolean {
  try {
    return bytesEqual(pubRawFromDid(did), pub);
  } catch {
    return false;
  }
}

export async function verifyB64url(
  pub: Uint8Array,
  sigB64url: string,
  message: Uint8Array,
): Promise<boolean> {
  try {
    const sig = b64urlToBytes(sigB64url);
    const key = await crypto.subtle.importKey("raw", pub, { name: "Ed25519" }, false, ["verify"]);
    return await crypto.subtle.verify({ name: "Ed25519" }, key, sig, message);
  } catch {
    return false;
  }
}

export async function importSigningKey(seed: Uint8Array): Promise<CryptoKey> {
  if (seed.length !== 32) throw new Error("ed25519 seed must be 32 bytes");
  const prefix = hexToBytes(PKCS8_PREFIX_HEX);
  const pkcs8 = new Uint8Array(prefix.length + seed.length);
  pkcs8.set(prefix, 0);
  pkcs8.set(seed, prefix.length);
  return crypto.subtle.importKey("pkcs8", pkcs8, { name: "Ed25519" }, false, ["sign"]);
}

export async function signB64url(key: CryptoKey, message: Uint8Array): Promise<string> {
  const sig = new Uint8Array(await crypto.subtle.sign({ name: "Ed25519" }, key, message));
  return bytesToB64url(sig);
}

export async function fingerprint(did: string): Promise<string> {
  return (await sha256Hex(did)).slice(0, 16);
}

export async function noteShardKey(did: string): Promise<[string, string]> {
  const fp = await fingerprint(did);
  return [fp.slice(0, 2), fp.slice(2)];
}

export async function didNoteNs(did: string): Promise<string> {
  const [shard] = await noteShardKey(did);
  return "did-" + shard;
}
